// @novonotes/webview-bridge/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Channel: () => Channel,
  default: () => index_default,
  detectEnvironment: () => detectEnvironment,
  invoke: () => invoke,
  webviewBridge: () => webviewBridge
});
module.exports = __toCommonJS(index_exports);
function detectEnvironment() {
  if (typeof window !== "undefined") {
    if ("__WXP_INTERNALS__" in window) {
      return "wxp";
    }
    if ("__TAURI__" in window || "__TAURI_INTERNALS__" in window) {
      return "tauri";
    }
  }
  return "unknown";
}
async function invoke(cmd, args) {
  const env = detectEnvironment();
  switch (env) {
    case "tauri": {
      const { invoke: tauriInvoke } = await import("@tauri-apps/api/core");
      return tauriInvoke(cmd, args);
    }
    case "wxp": {
      const windowWithInvoke = window;
      if (windowWithInvoke.invoke) {
        return windowWithInvoke.invoke(cmd, args);
      }
      throw new Error("wxp invoke not available");
    }
    default:
      throw new Error("No compatible WebView environment detected");
  }
}
var Channel = class {
  // private initializationPromise: Promise<void>;
  constructor(onmessage) {
    this._onmessage = onmessage;
    this.pendingId = `__CHANNEL__:pending_${Math.random().toString(36).substr(2, 9)}`;
    console.log("[webview-bridge] Channel constructor called");
    console.log("[webview-bridge] Environment detection:", {
      __TAURI__: "__TAURI__" in window,
      __TAURI_INTERNALS__: "__TAURI_INTERNALS__" in window,
      __WXP_INTERNALS__: "__WXP_INTERNALS__" in window,
      Channel: "Channel" in window
    });
    const env = detectEnvironment();
    console.log("[webview-bridge] Detected environment:", env);
    if (env === "tauri") {
      this.initializeTauriChannel();
    } else if (env === "wxp") {
      const wxpChannel = window.Channel;
      if (wxpChannel) {
        this.channel = new wxpChannel(this._onmessage);
        Promise.resolve();
      } else {
        this.initializeWxpChannel();
      }
    } else {
      this.initializeWithRetry();
    }
  }
  async initializeTauriChannel() {
    try {
      console.log("[webview-bridge] Initializing Tauri channel...");
      const { Channel: TauriChannel } = await import("@tauri-apps/api/core");
      console.log("[webview-bridge] TauriChannel imported successfully");
      this.channel = new TauriChannel();
      if (this._onmessage) {
        console.log(
          "[webview-bridge] Setting onmessage handler on Tauri channel"
        );
        this.channel.onmessage = this._onmessage;
      }
      console.log("[webview-bridge] Tauri channel created:", this.channel);
      console.log("[webview-bridge] Channel ID:", this.channel.id);
    } catch (error) {
      console.error(
        "[webview-bridge] Failed to initialize Tauri channel:",
        error
      );
    }
  }
  async initializeWxpChannel() {
    await new Promise((resolve) => setTimeout(resolve, 100));
    const retryChannel = window.Channel;
    if (retryChannel) {
      this.channel = new retryChannel(this._onmessage);
      console.log("[webview-bridge] wxp channel created after retry");
    } else {
      console.error("[webview-bridge] wxp Channel not available after retry");
    }
  }
  async initializeWithRetry() {
    await new Promise((resolve) => setTimeout(resolve, 100));
    const retryEnv = detectEnvironment();
    console.log("[webview-bridge] Retry environment detection:", retryEnv);
    if (retryEnv === "tauri") {
      await this.initializeTauriChannel();
    } else if (retryEnv === "wxp") {
      const wxpChannel = window.Channel;
      if (wxpChannel) {
        this.channel = new wxpChannel(this._onmessage);
        console.log("[webview-bridge] wxp channel created in retry");
      } else {
        await this.initializeWxpChannel();
      }
    } else {
      console.error(
        "[webview-bridge] No compatible WebView environment detected after retry"
      );
    }
  }
  get onmessage() {
    return this.channel?.onmessage;
  }
  set onmessage(handler) {
    this._onmessage = handler;
    if (this.channel) {
      console.log("[webview-bridge] Setting onmessage on existing channel");
      this.channel.onmessage = handler;
    } else {
      console.log(
        "[webview-bridge] Channel not yet initialized, storing onmessage for later"
      );
    }
  }
  // Tauriと同じtoJSON実装
  toJSON() {
    if (!this.channel) {
      console.warn(
        "[webview-bridge] toJSON called before channel initialization"
      );
      return this.pendingId;
    }
    if (typeof this.channel.toJSON === "function") {
      return this.channel.toJSON();
    }
    if (typeof this.channel.id === "number") {
      return `__CHANNEL__:${this.channel.id}`;
    }
    if (typeof this.channel.toIPC === "function") {
      return this.channel.toIPC();
    }
    throw new Error("Channel not properly initialized");
  }
};
var webviewBridge = {
  detectEnvironment,
  invoke,
  Channel
};
var index_default = webviewBridge;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Channel,
  detectEnvironment,
  invoke,
  webviewBridge
});
